src/fiber/config.rs

```diff
@@ -187,7 +214,10 @@ pub struct FiberConfig {
     pub(crate) sync_network_graph: Option<bool>,
 }
 
-#[derive(PartialEq, Copy, Clone, Default)]
+/// Must be a valid utf-8 string of length maximal length 32 bytes.
+/// If the length is less than 32 bytes, it will be padded with 0.
+/// If the length is more than 32 bytes, it should be truncated.
+#[derive(Eq, PartialEq, Copy, Clone, Default, Hash)]
```

Comment:
`Eq` and `Hash` is not necessary afer removing them from `BroadcastMessage`, as above.

Reply:
We used `HashSet` in https://github.com/contrun/fiber/blob/a7cb2bcfa0b84149c9f29539ac95cffc51b6e70b/src/fiber/gossip.rs#L995 to avoid saving the same `BroadcastMessageWithTimestamp`. This is the reason why I added `Eq` and `Hash` to `BroadcastMessageWithTimestamp` and related types.

Reply:
to resolve this review commment, suggest to implement PartialEq and Hash for BroadcastMessageWithTimestamp only:
```rust
#[derive(Debug, Clone, Eq)]
pub enum BroadcastMessageWithTimestamp {
    NodeAnnouncement(NodeAnnouncement),
    ChannelAnnouncement(u64, ChannelAnnouncement),
    ChannelUpdate(ChannelUpdate),
}

impl PartialEq for BroadcastMessageWithTimestamp {
    fn eq(&self, other: &Self) -> bool {
        self.cursor().eq(&other.cursor())
    }
}

impl Hash for BroadcastMessageWithTimestamp {
    fn hash<H: Hasher>(&self, state: &mut H) {
        self.cursor().hash(state)
    }
}
```

Reply:
Using only cursor to differentiate `BroadcastMessageWithTimestamp` is not enough. For example, an adversary may send us an invalid `ChannelUpdate` with the same timestamp and channel outpoint. As the `ChannelUpdate` can not be verified standalone (we need the corresponding `ChannelAnnouncement` to obtain the public key that should be used to verify the signature of `ChannelUpdate`). So if we are only using cursor as hash input, then it is possible the `ChannelUpdate` from a malicious user may replace a legitimate `ChannelUpdate` from an innocent user.

Reply:
there is a signature field, a malicious node cannot build a valid message for other nodes.

Reply:
For `ChannelAnnouncement` and `NodeAnnouncement`, that is true. The message definition of `ChannelUpdate` https://github.com/contrun/fiber/blob/aa8b783f2795eb5629d1c38bf38f3c1fec567dbf/src/fiber/schema/gossip.mol#L81-L95 contains only a signature without the corresponding public key. The public key for this signature should be looked up in the `ChannelAnnouncement` message. That means we can't verify it the signature of `ChannelUpdate` by the `ChannelUpdate` message itself. So it is possible for a malicious user to send us bogus `ChannelUpdate` message as long as we don't have the corresponding `ChannelAnnouncement`.

Reply:
I understand that the forged data will eventually fail the signature verification, it will not be stored and will not be re-broadcast to other nodes, and the node that forged it will be disconnected and banned, what is the issue with it?

Reply:
OK. Let me explain why do we need need a `Hash` implementation for `BroadcastMessageWithTimestamp` in the first place.
1. Upon receiving gossip messages, the node will first be saved a list `messages_to_be_saved` https://github.com/contrun/fiber/blob/e5f2a69f773f63202c807b7ae3172d997a2e50ac/src/fiber/gossip.rs#L995 in memory. This list serves two purposes. First, some messages may arrive out of order (with smaller timestamp), these messages need to be broadcasted to the network. We can't simply read latest cursor and broadcast messages after cursor to the network (this would miss some out of order messages). Second, some messages' dependencies are still not satisfied yet. So it is possible that we receive a `ChannelUpdate` without the corresponding `ChannelAnnouncement`. This list will buffer the incomplete messages.
2. The purpose of `Hash` for `BroadcastMessageWithTimestamp` is to avoid saving duplicated messages (we may receive duplicated gossip messages from different peers) to the list in memory (this list is actually a `HashSet`). This means if a message with the same hash is already saved from https://github.com/contrun/fiber/blob/e5f2a69f773f63202c807b7ae3172d997a2e50ac/src/fiber/gossip.rs#L1406 , then it wouldn't be inserted to the `messages_to_be_saved` again. So in the above scenario, if a malicious user saved a forged `ChannelUpdate` message to `messages_to_be_saved`, the legitimate `ChannelUpdate` we received will be ignored.

This is why `BroadcastMessageWIthTimestamp` needs a `Hash` implementation, and how using only `Cursor` may cause problems by a malicious user.